File: src/main/java/org/moinex/entities/CreditCardTransaction.java
/*
 * Filename: CreditCardTransaction.java
 * Created on: March  5, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.entities;

import jakarta.persistence.Column;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.MappedSuperclass;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import org.moinex.util.Constants;

/**
 * Base class for transactions
 */
@MappedSuperclass
@SuperBuilder
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public abstract class CreditCardTransaction
{
    @ManyToOne
    @JoinColumn(name = "crc_id", referencedColumnName = "id", nullable = false)
    private CreditCard creditCard;

    @Column(name = "date", nullable = false)
    private String date;

    @Column(name             = "amount",
            nullable         = false,
            scale            = 2)
    private BigDecimal amount;

    @Column(name = "description", nullable = true)
    private String description;

    public static abstract class CreditCardTransactionBuilder<
        C extends CreditCardTransaction, B extends CreditCardTransactionBuilder<C, B>>
    {
        public B date(LocalDateTime date)
        {
            this.date = date.format(Constants.DB_DATE_FORMATTER);
            return self();
        }
    }

    public LocalDateTime getDate()
    {
        return LocalDateTime.parse(date, Constants.DB_DATE_FORMATTER);
    }

    public void setDate(LocalDateTime date)
    {
        this.date = date.format(Constants.DB_DATE_FORMATTER);
    }
}

File: src/main/kotlin/org/moinex/model/creditcard/CreditCardTransaction.kt
/*
 * Filename: CreditCardTransaction.kt (original filename: CreditCardTransaction.java)
 * Created on: March  5, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 28/02/2026
 */

package org.moinex.model.creditcard

import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.MappedSuperclass
import org.moinex.common.converter.LocalDateTimeStringConverter
import org.moinex.common.toRounded
import java.math.BigDecimal
import java.time.LocalDateTime

@MappedSuperclass
class CreditCardTransaction(
    @ManyToOne
    @JoinColumn(name = "crc_id", referencedColumnName = "id", nullable = false)
    var creditCard: CreditCard,
    @Column(name = "date", nullable = false)
    @Convert(converter = LocalDateTimeStringConverter::class)
    var date: LocalDateTime,
    @Column(name = "amount", nullable = false, scale = 2)
    var amount: BigDecimal,
    @Column(name = "description")
    var description: String?,
) {
    init {
        amount = amount.toRounded()
        require(amount > BigDecimal.ZERO) { "Amount must be positive" }
    }
}
